custom-targets/vertex-ai/model-deployer/deploy.go

```diff
@@ -0,0 +1,211 @@
+	d.addCommonMetadata(res)
+
+	fmt.Println("Uploading successful deploy results")
+	rURI, err := d.req.UploadResult(ctx, d.gcsClient, res)
+	if err != nil {
+		return fmt.Errorf("error uploading deploy results: %v", err)
+	}
+	fmt.Printf("Uploaded deploy results to %s\n", rURI)
+	return nil
+
+}
+
+// deploy performs the Vertex AI model deployment
+func (d *deployer) deploy(ctx context.Context) (*clouddeploy.DeployResult, error) {
+
+	if err := d.downloadManifest(ctx); err != nil {
+		return nil, err
+	}
+
+	manifestData, err := d.applyModel(ctx, localManifest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to deploy model: %v", err)
+	}
+
+	mURI, err := d.req.UploadArtifact(ctx, d.gcsClient, "manifest.yaml", &clouddeploy.GCSUploadContent{Data: manifestData})
+	if err != nil {
+		return nil, fmt.Errorf("error uploading deploy artifact: %v", err)
+	}
+
+	return &clouddeploy.DeployResult{
+		ResultStatus:  clouddeploy.DeploySucceeded,
+		ArtifactFiles: []string{mURI},
+	}, nil
+}
+
+// downloadManifest downloads the rendered manifest from Google Cloud Storage to the local manifest file path
+func (d *deployer) downloadManifest(ctx context.Context) error {
+	fmt.Printf("Downloading deploy input manifest from %q.\n", d.req.ManifestGCSPath)
+
+	downloadPath, err := d.req.DownloadManifest(ctx, d.gcsClient, localManifest)
+	if err != nil {
+		fmt.Printf("Unable to download deployed manifest from: %s.\n", d.req.ManifestGCSPath)
+		return fmt.Errorf("unable to download deploy input from %s: %v", d.req.ManifestGCSPath, err)
+	}
+
+	fmt.Printf("Downloaded deploy input manifest from: %s\n", downloadPath)
+
+	return nil
+}
+
+// addCommonMetadata inserts metadata into the deploy result that should be present
+// regardless of deploy success or failure.
+func (d *deployer) addCommonMetadata(rs *clouddeploy.DeployResult) {
+	if rs.Metadata == nil {
+		rs.Metadata = map[string]string{}
+	}
+	rs.Metadata[clouddeploy.CustomTargetSourceMetadataKey] = aiDeployerSampleName
+	rs.Metadata[clouddeploy.CustomTargetSourceSHAMetadataKey] = clouddeploy.GitCommit
+}
+
+// applyModel deploys the DeployModelRequest parsed from `localManifest`
+// it returns the DeployedModelRequest object that was used in yaml format.
+func (d *deployer) applyModel(ctx context.Context, localManifest string) ([]byte, error) {
+
+	deployModelRequest, err := deployModelFromManifest(localManifest)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load DeployModelRequest from manifest: %v", err)
+	}
+
+	if d.req.Percentage != 100 {
+		if err := d.makeManifestChangesForCanary(deployModelRequest); err != nil {
+			return nil, fmt.Errorf("unable to make canary changes to the manifest: %v", err)
+		}
+	}
+
+	op, err := d.aiPlatformService.Projects.Locations.Endpoints.DeployModel(d.params.endpoint, deployModelRequest).Do()
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to deploy model: %v", err)
+	}
+
+	if err := poll(ctx, d.aiPlatformService, op); err != nil {
+		return nil, err
+	}
```

Comment:
Move this logic to deployModel(...) function in `vertexai.go`. The benefit to introduce that file with helper functions is that we can get rid of things like op polling from here so user can just see ah this deploys the model in a single line hiding away the poll logic.